File: solution/448FindAllNumbersDisappearedInAnArray/448FindAllNumbersDisappearedInAnArray.cpp
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> findDisappearedNumbers(vector<int> &nums) {
    vector<int> result;
    for (auto &x : nums) {
      --x;
    }
    int i = 0;
    while (i < nums.size()) {
      if (i != nums[i]) {
        if (nums[nums[i]] == nums[i]) {
          nums[i] = -1;
        } else {
          swap(nums[i], nums[nums[i]]);
        }
        if (nums[i] == -1) {
          ++i;
        }
      } else {
        ++i;
      }
    }
    for (i = 0; i < nums.size(); i++) {
      if (nums[i] == -1) {
        result.push_back(i + 1);
      }
    }
    return result;
  }
};

int main() { return 0; }

File: src/448FindAllNumbersDisappearedInAnArray/448FindAllNumbersDisappearedInAnArray.rs
impl Solution {
    pub fn find_disappeared_numbers(nums: Vec<i32>) -> Vec<i32> {
        let mut nums: Vec<i32> = nums.into_iter().map(|x| x-1).collect();
        let n = nums.len();
        let mut p = 0 as usize;
        while p < n {
            if nums[p] != p as i32 {
                let q = nums[p] as usize;
                if nums[q] != q as i32 {
                    nums.swap(p, q);
                    continue;
                }
            }    
            p += 1;
        }
        
        let mut result = vec![];
        for p in 0..n {
            if nums[p] != p as i32 {
                result.push( (p+1) as i32);
            }
        }
        
        result
    }
}
